contracts/RMRK/extension/erc20Holder/ERC20Holder.sol

```diff
@@ -19,32 +19,42 @@ abstract contract ERC20Holder is IERC20Holder {
         address erc20Contract,
         uint256 tokenId
     ) external view returns (uint256) {
-        return balances[tokenId][erc20Contract];
+        return _balances[tokenId][erc20Contract];
     }
 
     /**
-     * @inheritdoc IERC20Holder
+     * @notice Transfer ERC-20 tokens to a specific token
+     * @dev The ERC-20 contract must have approved this contract to transfer the ERC-20 tokens
+     * @dev The balance MUST be transferred from the message sender
+     * @dev Implementers should validate that the msg sender is either the token owner or approved before calling this.
+     * @param erc20Contract The ERC-20 contract
+     * @param tokenId The token to transfer to
+     * @param value The number of ERC-20 tokens to transfer
+     * @param data Additional data with no specified format, to allow for custom logic.
      */
-    function transferERC20FromToken(
+    function _transferERC20FromToken(
```

Comment:
Should this also accept `destinationId` to be able to transfer the tokens from one NFT to another?

Reply:
Did not think of this actually, it makes it more complex. But would be good to have. 
I don't like that we need an additional flag to know if destination is an NFT or not, since Id could be zero.  
I would prefer keeping it simply, WDTY?

Reply:
I get the simplicity. Just thinking about the most usability? We've been dealing with simplicity vs. usability for all of the proposals we published..

Reply:
I've spent some time thinking about it and found a blocker. Remember we cannot push ERC20 tokens as there is no onERC20Received method. 
However, I see an alternative which would be to transfer ERC20s between NFTs of the same collection, it's more of a util, will add a method and let's see if it's worth it.

Reply:
@ThunderDeliverer  I could not find a way to do this in a secure way.
I evaluated 2 options, say token 1 from collection A wants to send ERC20 tokens X to token 2 from collection B. Token A is owned by User.

Option A. Pulling.
1. User  needs to approve B to mange it's token. So the user calls `A.approve(B, 1)`.
2. Owner calls `B.receiveERC20FromToken(fromCollection: A, fromToken: 1, toToken 2, amount: x).`  
3. B would in turn call `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, amount: x)`
4. B would include the amount x in balances held by 2.

This has a flaw, after step 1, anyone could do step 2 and transfer to theirs. Unless it checks that msg.sender owns A.1. Which could be done, but it's scary to say the least. A lot of trust has to be put on the receiver (B).

Also, the fact that you have to approve the destination contract is very disturbing, it would still be your fault but it makes it easier to fall for scams IMO.

Option B. Pushing (usually terrible idea, but let's see).
1.  User calls `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, toTokenId: 2, amount: x)`
2.  A, after doing the transfer, notifies B that the funds it received where for token 2. 

The problem here is more obvious, A can lie.

Option C. Pushing V2
1.  User calls `A.transferERC20FromToken(erc20: X, fromToken: 1, toCollection: B, toTokenId: 2, amount: x)`
2. A approves B to spend X:  `X.approve(spender: B, amount: x)`
3. A calls `B.transferERC20ToToken(erc20: X, tokenId: 2, amount: x, from: A)` Notices that we need to add `from`, so msg.sender is not necessarily used.
4. A reduces balance of X from 1.

The problem here is that anyone could call step 3 with a different from that has some pending approval. 

I don't like option A, and would never do B or C. I consider the safest is to simply not do it, and if the transfer between tokens method makes it look limiting, I would simply remove it and have sources and destination be only addresses, not NFTs.

If you want to give it a try go ahead, but I'd rather have us focused on other priorities, this is perfectly good enough for the use case I've described in the past.

Reply:
I'll leave my thoughts on the possible approach and we can either consider them or remove the ability to transfer ERC20s between tokens.

The ERC20 smart contract considers smart contract A the owner of the tokens, and the User is only the owner of the NFT residing in A.
I'd add 2 functions; `sendFromTokenToToken` and `receiveFromToken` (names applicable to change 😅 ).
So, the User calls `sendFromTokenToToken` and the following happens (in the same transaction):
1. A calls ERC20 contract to grant the allowance for B
2. A calls B telling it how many tokens to transfer to a token
3. B transfers the ERC20s to token (the NFT to receive must exist)

I think this is it. No danger of another token to steal the ERC20s. The only danger is the possibility of malicious B, but this is a permanent vulnerability that can't be removed 🤔 

PS.: Also, the EIP wording will be a challenge as it can get confusing really fast 😂

Reply:
Could work, wanna give it a try?